### DISCUSSION TITLE
Audio and MIDI network protocols used?

### DISCUSSION THREAD

**Discussion by celoranta** [apohl79/audiogridder#725] [General]
Hi all.

I’m curious what AG uses “under the hood” to facilitate real-time audio and midi transport over LAN.  I can hazard a guess that OSC is involved on the midi side, but I’m stumped regarding the audio.  There are no network hardware requirements as per AVB, and obviously nobody is paying for a Dante license here.  AES67, perhaps?  So curious!

**Comment by apohl79** [comment]
AG uses a custom TCP protocol for audio/midi as of today. I've always planned to implement audio/midi streaming via UDP based RTP but that takes time I did not have yet. But as it is based on standard TCP right now, there is no need for special hardware.

**Comment by celoranta** [comment]
Very very cool. 
There are very few options available for real time multichannel audio over Ethernet streaming.  Especially in the open source community.  Very well done!

I’d like to create a Juce plug-in Virtual Instrument which acts as a digital multichannel and midi output from the hex pickup and OSC controls on my guitar.  (Plug the guitar into Ethernet, insert the plugin into my performance DAW, and six channels of audio plus midi CC messages emerge from the plugin in real time.)  Basically a virtual instrument that outputs the digitized signal from a physical instrument.  

You’ve obviously solved all the problems I’d need to address during your dev of AG.  Can you point me to any resources that would help me get there?

Best,

Chris

**Reply by apohl79** [reply]
Is that similar to oscvstbridge? I have not worked with OSC, but as it is open, there might be usable tools around this. So not sure how the protocol is designed. In the end you will have to make your plugin receive OSC messages, unpack the midi data and inject it into the midi buffer passed to your audio processors processBlock method.

**Reply by celoranta** [reply]
Thanks!

I’m less concerned about midi/OSC, and more concerned about writing a custom tcp audio transport protocol (likely on an xmos xcore200) and accessing it through Juce.  Any advice there?

**Comment by celoranta** [comment]
Andreas,

I haven't used OSCVST bridge previously, but yes, I'd be looking for
something similar.  I'm feeling more confident about the OSC/MIDI side, as
they aren't nearly as real-time dependent as uncompressed digital audio
streams.  I would probably configure one core of an XMOS XCORE200 to
translate digital sensor inputs to values and send them over Ethernet.  OSC
just happens to be a pre-existing network transport medium for MIDI data,
so I'd look there first.
As for the audio, I'd be using a Cirrus Logic ADC to convert up to 16
channels of pickup output to 2 x digital TDM8 streams, routing the streams
to the XCORE, and using its I2S/TDM library to send the data over an
ethernet server socket via the I2S/TDM library's callback functions.
That's really the extent of what I understand, though.
I do think I'll have the XCore run separate web services for Audio and
OSC/MIDI... technically the audio is bound for the Juce plugin while the
OSC is bound for integration with the VST host software.  Separating may be
best?
Let me know if you have any pointers on creating the actual custom digital
audio ethernet transport protocol layer.  That's voodoo to me thus far.

Chris

**Reply by apohl79** [reply]
Well not really. I've really just put things together from scratch. In AG's case also, there is no conversions happening from one protocol to another. It is basically just serializing the JUCE plugin interface data to TCP to bridge it over the wire. On the server side, where the remote plugins are hosted, the data is deserialized 1-1 and passed to the plugins. No woodoo. :-)

I think I don't fully understand your intended data flows yet. If you have access to the raw audio channel data, you can serialize this directly to the network, like AG does. This is not super efficient though, but relatively easy to implement. Look here to see how this happens today: https://github.com/apohl79/audiogridder/blob/524a008774c8455ea921274a326c200929af8a10/Common/Source/Message.hpp#L154
